### ISSUE TITLE
Implementation of "FastMaths" functions in ASM

### ISSUE THREAD

**Issue by jdelauney** [jdelauney/SIMD-VectorMath-UnitTest#13]
During my research over the web i found some interesting discussions or articles about fast sin/cos .... functions what do you think about ?

**Comment by jdelauney** [comment]
Some links i've found  
- http://forum.devmaster.net/t/fast-and-accurate-sine-cosine/9648
- https://dtosoftware.wordpress.com/2013/01/07/fast-sin-and-cos-functions/
- https://stackoverflow.com/questions/18662261/fastest-implementation-of-sine-cosine-and-square-root-in-c-doesnt-need-to-b
- http://forum.devmaster.net/t/fast-and-accurate-sine-cosine/9648/28
-http://www.netlib.org/cephes/

**Comment by dicepd** [comment]
Interesting read but I very rarely use sin or cos direct anymore as I have gotten into using unit vector or whatever,  for just about everything I do, calculated once as I build whatever it is I need. A unit vector is basically the cos angles, sin is 1-cos and I avoid Tan like the plague. When it gets transformed I can read new world coordinate cos of most things from an attached unit vector. I now tend to think if I am having to use sin or cos direct then I am not doing it right, rethink the problem and try to make them go away, I usually end up with more efficient code.

But as we are using  xmm regs for most floating point ops and we wish to reduce or limit the fpu calls I would go for one of the pick your own resolution methods over Taylor any day.

**Comment by dicepd** [comment]
https://www.dsprelated.com/showarticle/1052.php

fast atan2 algo I came across while looking for something else. No idea how it compares to what we have at the momemt, maybe worth a look.

**Comment by jdelauney** [comment]
Just added FastMath base unit. You'll find some interesting links in the header of GLZFastMath.pas
I've also made a "quick and dirty" little test App

**Comment by dicepd** [comment]
https://stackoverflow.com/questions/13771573/yet-another-fast-trigonometry

Check out the post further down by Joseph Quinsey where he is using integer math in C looks to be a very interesting approach from years gone by.

**Comment by dicepd** [comment]
Suggestion for remez 
```pascal
type
{$ifdef REMEZ_DOUBLE}
  TSinCos = record
    sin: double;
    cos: double;
  end;
  PSinCos = ^TSinCos;
{$else}
  TSinCos = record
    sin: single;
    cos: single;
  end;
  PSinCos = ^TSinCos;
{$endif}

  TRemezArrayType = array[0..4] of TSinCos;    

implementation 
 remez_const : TRemezArrayType = ((sin:0.99997937679290771484375;
                                   cos:1.0),
                                  (sin:-0.166624367237091064453125;
                                   cos:-0.49985158205212948 ),
                                  (sin:8.30897875130176544189453125e-3;
                                   cos:4.151803521644389800e-2 ),
                                  (sin:-1.92649182281456887722015380859375e-4;
                                   cos:-1.34229470252267675e-3 ),
                                  (sin:2.147840177713078446686267852783203125e-6 ;
                                   cos:1.89298648243010005378029e-5 ));        
```
save getting lots of consts mixed up in asm code and gives us a single pointer to load with offfsets for SSE, and we can test the speed of pushing to double and seeing if that improve accuracy.

**Comment by jdelauney** [comment]
Yes using Fixed point can be a great optimization in computing. But we must ensure the op for integer is fastest than float else  it's not necessary and interesting for us

**Comment by jdelauney** [comment]
Regarless to my post yesterday about the sin function in issue #29 
i come back to old school. Now sin/cos use a LUT more accuracy and fastest

![2018-02-13_162512](https://user-images.githubusercontent.com/29355534/36157795-ef66e0f8-10da-11e8-8431-49cb189a7546.jpg)

the average FPS is around 15/16 not too bad for a software raymarching with phong illumination

**Comment by dicepd** [comment]
Not bad at all, as one who can remember waiting 5 mins for AutoCad to phong render a svga sized window, it shows how far computing has come.
